<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<title>RRBS-Kit</title>
<style type="text/css">
<!--
body {
	font: 100%/1.4 Verdana, Arial, Helvetica, sans-serif;
	background: #4E5869;
	margin: 0;
	padding: 0;
	color: #000;
}
ul, ol, dl {
	padding: 0;
	margin: 0;
}
h1, h2, h3, h4, h5, h6, p {
	margin-top: 0;
	padding-right: 15px;
	padding-left: 15px;
}
a img {
	border: none;
}
a:link {
	color:#414958;
	text-decoration: underline;
}
a:visited {
	color: #4E5869;
	text-decoration: underline;
}
a:hover, a:active, a:focus {
	text-decoration: none;
}
.container {
	width: 80%;
	max-width: 1260px;
	min-width: 780px;
	background: #FFF;
	margin: 0 auto;
}
.header {
	background: #6F7D94;
}
.content {
	padding: 10px 0;
}
.content ul, .content ol {
	padding: 0 15px 15px 40px;
}
.footer {
	padding: 10px 0;
	background: #6F7D94;
}
.fltrt {
	float: right;
	margin-left: 8px;
}
.fltlft {
	float: left;
	margin-right: 8px;
}
.clearfloat {
	clear:both;
	height:0;
	font-size: 1px;
	line-height: 0px;
}
-->
</style>
</head>

<body>
<div class="container">
  <div class="header"><a href="#"><img src="image/logo.png" alt="" name="Insert_logo" width="20%" height="90" id="Insert_logo" style="background: #8090AB; display:block;" /></a> 
    <!-- end .header --></div>
  <div class="content">
    <h1>RRBS-Kit</h1>
    <p>URL: <a href=https://code.google.com/p/rrbs-kit/source/browse/>https://code.google.com/p/rrbs-kit/source/browse/</a></p>
    <h2>Introduction</h2>
    <p>Typically, RRBS samples are generated by digesting genomic DNA with the restriction endonuclease MspI. This is followed by insert-size select (50-220bp), ending repair, A-tailing, adapter ligation and finally bisulfite conversion. Often, the library is also size-selected for fragments between 40 and 220bp in length, but we just use 50-220bp fragments in our work flow. This fragment size has been shown to be plentiful in the sample and yield information on the vast majority of CpG islands (CGIs) in the human or mouse genome. The below is a simple flowchart for RRBS.</p>
    <div align="center"><img src="image/RRBS.png" width="60%" height="60%" alt="RRBS"> </div>
    <h2>Workflow</h2>
    <p>We list main processing step for RRBS data. Actually, we do not show the whole detailed step and QC insertion points. In our workflow, we used an ingenious self-check mechanism to ensure processing done successfullly and easily used. These modules are called Jobcluster© and FlowSmart©. And they make the flow continuous, exact and end to end.</p>
    <div align="center"><img src="image/workflow.png" width="60%" height="60%" alt="workflow"> </div>
    <h2>File Format</h2>
    <h3>Fastq file format</h3>
    <p>A FASTQ file normally uses four lines per sequence. Line 1 begins with a '@' character and is followed by a sequence identifier and an optional description (like a FASTA title line). Line 2 is the raw sequence letters. Line 3 begins with a '+' character and is optionally followed by the same sequence identifier (and any description) again. Line 4 encodes the quality values for the sequence in Line 2, and must contain the same number of symbols as letters in the sequence. The base's quality is equal to ASCII value in line that start from 33 (Sanger FASTQ-like format) / 64 (Illumina 1.3+ FASTQ-like format). Starting from the Illumina GA Pipeline v1.7, the range of base quality values is from 2 to 41 (66 - 105).</p>
    <div align="center"><img src="image/fastq.png" width="90%" height="90%" alt="fastq"> </div>
    <h3>Variant call format</h3>
    <p>For detailed, click the link `1000 Genomes VCF(Variant Call Format) version 4.1'.</p>
    <p>Reference: <a href=http://www.1000genomes.org/wiki/Analysis/Variant%20Call%20Format/vcf-variant-call-format-version-41>http://www.1000genomes.org/wiki/.../vcf-variant-call-format-version-41</a></p>
    <h3>DMR format</h3>
    <p>In order to accurately and clearly describe the DMR(differentially methylated regions), we defined *.DMR text file format. It contains meta-information lines, a header line, and then data lines each containing information about a continuous region in the genome. The following is a detailed description of the DMR file format.</p>
    <p>Reference: <a href=image/Logo.png>Logo</a></p>
    <h2>Dataset & Tools</h2>
	<h3>Human reference genome</h3>
    <p>hg18/NCBI Build36 
    <br>Download from <a href=http://hgdownload.cse.ucsc.edu/goldenPath/hg18/bigZips/chromFa.zip>http://hgdownload.cse.ucsc.edu/goldenPath/hg18/bigZips/chromFa.zip</a>.
    hg19/NCBI Build37
    <br>Download from <a href=http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz">http://hgdownload.cse.ucsc.edu/goldenPath/hg19/bigZips/chromFa.tar.gz</a>.
    <br>Note: The chromosome coordinates used in our whole analysis corresponded to UCSC Santa Cruz hg18/hg19.
    </p>
    <h3>Mapping software</h3>
    <p>bsmap<br>
      Download from <a href=http://soap.genomics.org.cn/soapaligner.html> http://code.google.com/p/bsmap/.</a>
   <br>
   bowtie2
   <br>Download from <a href=http://bowtie-bio.sourceforge.net/bowtie2/index.shtml> http://bowtie-bio.sourceforge.net/bowtie2/index.shtml.</a> 
    </p>
    <h3>Other Bam/Sam utils</h3>
    <p>
    picard 
    <br>Download from <a href=http://picard.sourceforge.net/index.shtml> http://picard.sourceforge.net/index.shtml.</a>
    </p>
    <h2>Analysis Results</h2>
    <p>Please click the link below to view the main results.
    <br><a href="./report">report</a>
    </p>
    <h2>Contact information</h2>
    <p>If you have any problem please do not hesitate to contact us!
    <br><a href="mailto:gaoshengjie@genomics.cn?subject=bug report&cc=zoudan@genomics.cn&body=%0d%0aPlease send us your operator system version, used parameters or/and error infomation when you 'feel' something wrong happened.%0d%0a">gaoshengjie@genomics.cn</a>
    </p>
    <!-- end .content --> 
  </div>
  <div class="footer">
    <p></p>
    <!-- end .footer --></div>
  <!-- end .container --> 
</div>
</body>
</html>
